# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: dom/media/systemservices/MediaUtils.cpp
# Commit: c34ea4d71235
# Full Hash: c34ea4d71235c0519020893ba04d7206e7fe4f14
# Author: Michael Froman <mfroman@mozilla.com>
# Date: 2021-11-01 18:17:33
# Regressor Bug: 1654112
# File Overlap Count: 1
# Description:
#   Bug 1654112 - fix tsan error, using a destroyed mutex. r=ng
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D121007
# ==============================================================================

diff -r f547e3673690 -r c34ea4d71235 dom/media/systemservices/MediaUtils.cpp
--- a/dom/media/systemservices/MediaUtils.cpp	Wed Jul 21 07:51:58 2021 -0700
+++ b/dom/media/systemservices/MediaUtils.cpp	Mon Jul 26 16:42:40 2021 -0500
@@ -56,31 +56,32 @@
 
 class RefCountedTicket {
   RefPtr<MediaEventBlocker> mBlocker;
-  MediaEventForwarder<void> mShutdownEventForwarder;
+  RefPtr<MediaEventForwarder<void> > mShutdownEventForwarder;
 
  public:
   NS_INLINE_DECL_THREADSAFE_REFCOUNTING(RefCountedTicket)
 
   RefCountedTicket()
-      : mShutdownEventForwarder(GetMainThreadSerialEventTarget()) {}
+      : mShutdownEventForwarder(
+            new MediaEventForwarder<void>(GetMainThreadSerialEventTarget())) {}
 
   void AddBlocker(nsString aName, nsString aFileName, int32_t aLineNr) {
     MOZ_ASSERT(NS_IsMainThread());
     MOZ_ASSERT(!mBlocker);
     mBlocker = MakeAndAddRef<MediaEventBlocker>(aName);
-    mShutdownEventForwarder.Forward(mBlocker->ShutdownEvent());
+    mShutdownEventForwarder->Forward(mBlocker->ShutdownEvent());
     GetShutdownBarrier()->AddBlocker(mBlocker.get(), std::move(aFileName),
                                      aLineNr, std::move(aName));
   }
 
-  MediaEventSource<void>& ShutdownEvent() { return mShutdownEventForwarder; }
+  MediaEventSource<void>& ShutdownEvent() { return *mShutdownEventForwarder; }
 
  protected:
   virtual ~RefCountedTicket() {
     MOZ_ASSERT(NS_IsMainThread());
     MOZ_ASSERT(mBlocker);
     GetShutdownBarrier()->RemoveBlocker(mBlocker.get());
-    mShutdownEventForwarder.DisconnectAll();
+    mShutdownEventForwarder->DisconnectAll();
   }
 };
 }  // namespace